File: src/main/java/io/github/bot/terminal/application/domains/monitoring/rest/MonitoringTasksRestService.java
package io.github.bot.terminal.application.domains.monitoring.rest;

import io.github.bot.terminal.application.domains.bot_commands.rest.dto.BotCommandTypeDTO;
import io.github.bot.terminal.application.domains.common.action.values.ActionType;
import io.github.bot.terminal.application.domains.common.conditions.dto.ConditionTypeDTO;
import io.github.bot.terminal.application.domains.common.conditions.values.ConditionType;
import io.github.bot.terminal.application.domains.common.cron.dto.CronFrequencyDTO;
import io.github.bot.terminal.application.domains.common.cron.values.Cron;
import io.github.bot.terminal.application.domains.monitoring.repository.MonitoringTaskDetails;
import io.github.bot.terminal.application.domains.monitoring.repository.MonitoringTaskRepository;
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskDTO;
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskTypeDTO;
import io.github.bot.terminal.application.domains.monitoring.rest.requests.MonitoringTaskRequest;
import io.github.bot.terminal.application.domains.workers.MonitoringTasksWorker;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
@RequiredArgsConstructor
public class MonitoringTasksRestService {

    private final MonitoringTaskRepository repository;
    private final MonitoringTasksRestConverter converter;
    private final MonitoringTasksWorker worker;

    public MonitoringTaskDTO add(MonitoringTaskRequest request) {
        MonitoringTaskDetails details = converter.mapToDetails(request);
        repository.add(details);
        worker.runRefreshSchedulers();
        return converter.mapToDto(details);
    }

    public MonitoringTaskDTO edit(String id, MonitoringTaskRequest request) {
        MonitoringTaskDetails detailsUpdate = converter.mapToDetails(request);
        MonitoringTaskDetails details = getById(id);
        detailsUpdate.setId(details.getId());
        repository.update(detailsUpdate);
        worker.runRefreshSchedulers();
        return converter.mapToDto(detailsUpdate);
    }

    public List<MonitoringTaskDTO> list() {
        return repository.findAll()
                .stream().map(converter::mapToDto)
                .collect(Collectors.toList());
    }

    public MonitoringTaskDTO get(String id) {
        return converter.mapToDto(getById(id));
    }

    public void delete(String id) {
        repository.deleteById(id);
        worker.runRefreshSchedulers();
    }

    private MonitoringTaskDetails getById(String id) {
        return repository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Monitoring task not found: id=" + id));
    }

    public List<CronFrequencyDTO> cronFrequencies() {
        return Arrays.stream(Cron.values())
                .map(t -> new CronFrequencyDTO().setLabel(t.getLabel()).setCron(t.getCron()))
                .collect(Collectors.toList());
    }

    public List<MonitoringTaskTypeDTO> types() {
        return Stream.of(ActionType.SQL_SELECT_AS_ONE_NUMBER)
                .map(t -> new MonitoringTaskTypeDTO().setLabel(t.getLabel()).setType(t.name()))
                .collect(Collectors.toList());
    }

    public List<ConditionTypeDTO> conditionTypes() {
        return Stream.of(ConditionType.values())
            .map(t -> new ConditionTypeDTO().setLabel(t.getLabel()).setType(t.name()))
            .collect(Collectors.toList());
    }
}

File: src/main/java/io/github/bot/terminal/application/domains/monitoring/rest/MonitoringTasksRestService.kt
package io.github.bot.terminal.application.domains.monitoring.rest

import io.github.bot.terminal.application.domains.common.action.values.ActionType
import io.github.bot.terminal.application.domains.common.conditions.dto.ConditionTypeDTO
import io.github.bot.terminal.application.domains.common.conditions.values.ConditionType
import io.github.bot.terminal.application.domains.common.cron.dto.CronFrequencyDTO
import io.github.bot.terminal.application.domains.common.cron.values.Cron
import io.github.bot.terminal.application.domains.monitoring.repository.MonitoringTaskDetails
import io.github.bot.terminal.application.domains.monitoring.repository.MonitoringTaskRepository
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskDTO
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskTypeDTO
import io.github.bot.terminal.application.domains.monitoring.rest.requests.MonitoringTaskRequest
import io.github.bot.terminal.application.domains.workers.MonitoringTasksWorker
import org.springframework.stereotype.Service
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream

@Service
class MonitoringTasksRestService(
        private val repository: MonitoringTaskRepository,
        private val converter: MonitoringTasksRestConverter,
        private val worker: MonitoringTasksWorker
) {

    fun add(request: MonitoringTaskRequest): MonitoringTaskDTO {
        val details = converter.mapToDetails(request)
        repository.add(details)
        worker.runRefreshSchedulers()
        return converter.mapToDto(details)
    }

    fun edit(id: String, request: MonitoringTaskRequest): MonitoringTaskDTO {
        val detailsUpdate = converter.mapToDetails(request)
        val details = getById(id)
        details.merge(detailsUpdate)
        repository.update(details)
        worker.runRefreshSchedulers()
        return converter.mapToDto(details)
    }

    fun list(): List<MonitoringTaskDTO> {
        return repository.findAll()
                .stream().map { details: MonitoringTaskDetails -> converter.mapToDto(details) }
                .collect(Collectors.toList())
    }

    operator fun get(id: String): MonitoringTaskDTO {
        return converter.mapToDto(getById(id))
    }

    fun delete(id: String) {
        repository.deleteById(id)
        worker.runRefreshSchedulers()
    }

    private fun getById(id: String): MonitoringTaskDetails {
        return repository.findById(id)
                ?: throw IllegalArgumentException("Monitoring task not found: id=$id")
    }

    fun cronFrequencies(): List<CronFrequencyDTO> {
        return Arrays.stream(Cron.values())
                .map { t: Cron -> CronFrequencyDTO(t.label,t.cron) }
                .collect(Collectors.toList())
    }

    fun types(): List<MonitoringTaskTypeDTO> {
        return Stream.of(ActionType.SQL_SELECT_AS_ONE_NUMBER)
                .map { MonitoringTaskTypeDTO(label = it.label, type = it.name) }
                .collect(Collectors.toList())
    }

    fun conditionTypes(): List<ConditionTypeDTO> {
        return Stream.of(*ConditionType.values())
                .map { t: ConditionType -> ConditionTypeDTO(t.label, t.name) }
                .collect(Collectors.toList())
    }
}